platform/android/src/style/sources/geojson_source.cpp

```diff
@@ -125,6 +110,55 @@ namespace android {
         return jni::Object<Source>(GeoJSONSource::javaClass.New(env, constructor, reinterpret_cast<jni::jlong>(this)).Get());
     }
 
+    template <class JNIType>
+    void GeoJSONSource::setCollectionAsync(jni::JNIEnv& env, jni::Object<JNIType> jObject) {
+
+        std::shared_ptr<jni::jobject> object = std::shared_ptr<jni::jobject>(jObject.NewGlobalRef(env).release()->Get(), GenericGlobalRefDeleter());
+
+        // creating a new Update
+        std::unique_ptr<CollectionUpdate> collectionUpdate = makeUpdate<CollectionUpdate>(object);
+        collectionUpdate->makeUpdateFn<JNIType>();
+
+        setAsync(std::move(collectionUpdate));
+    }
+
+    template <class UpdateType, typename ObjectType>
+    std::unique_ptr<UpdateType> GeoJSONSource::makeUpdate(ObjectType object) {
+        // creating a new Update
+        return std::make_unique<UpdateType>(object, std::make_unique<Actor<Callback>>(
+                *Scheduler::GetCurrent(),
+                [this](GeoJSON geoJSON) {
+                    // conversion from Java features to core ones finished
+                    android::UniqueEnv _env = android::AttachEnv();
+
+                    // Update the core source
+                    source.as<mbgl::style::GeoJSONSource>()->GeoJSONSource::setGeoJSON(geoJSON);
+
+                    // if there is an awaiting update, execute it, otherwise, release resources
+                    if (awaitingUpdate) {
+                        update = std::move(awaitingUpdate);
+                        awaitingUpdate.reset();
+                        update->updateFn(converter->self());
+                    } else {
+                        update.reset();
+                    }
+                }));
+    }
+
+    void GeoJSONSource::setAsync(std::unique_ptr<Update> queuedUpdate) {
+        awaitingUpdate = std::move(queuedUpdate);
+
+        // If another update is running, wait
+        if (update) {
+            return;
+        }
+
+        // no updates are being process, execute this one
+        update = std::move(awaitingUpdate);
+        awaitingUpdate.reset();
```

Comment:
This `reset` should be unnecessary because the move assignment will have already emptied `awaitingUpdate`.